File: Python/19.RemoveNthNodeFromEndOfList.py
"""
This problem showed me the data structure in Python
is somewhat different from what I fimiliar with in C++
OOP
"""
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head, n):
        dummy = ListNode(0, head)
        slow = dummy
        fast = dummy
        
        # move fast pointer n steps ahead
        for _ in range(n+1):
            fast = fast.next
            
        # move both pointers until fast reaches the end
        while fast:
            slow = slow.next
            fast = fast.next
            
        # remove the nth node from the end
        slow.next = slow.next.next
        return dummy.next

File: C/19.RemoveNthNodeFromEndOfList.c
//The application of dummy is to handle the edge case when the head node is removed.
 struct ListNode {
  int val;
  struct ListNode *next;
  };

struct ListNode* removeNthFromEnd(struct ListNode* head, int n) {
    struct ListNode *fast, *slow;
    struct ListNode* dummy=(struct ListNode*)malloc(sizeof(struct ListNode));
    dummy->next=head;
    fast=dummy;
    slow=dummy;
    for(int i=0; i<n+1; i++){
        fast=fast->next;
    }
    while(fast){
        slow=slow->next;
        fast=fast->next;
    }
    slow->next=slow->next->next;
    return dummy->next;
}